### ISSUE TITLE
[schedv2] - Bug: Lapses in Filtered Decks use 'new' delay steps

### ISSUE THREAD

**Issue by david-allison-1** [ankidroid/Anki-Android#6082]
###### Reproduction Steps

Reported by /u/CarelessSecretary9 on reddit for 2.9.6. Confirmed on 2.10alpha71

Relevant settings dump (via DataGrip)

<details><summary>Table Data</summary>

| id | nid | did | ord | mod | usn | type | queue | due | ivl | factor | reps | lapses | left | odue | odid | flags | data |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1510928829863 | 1510928805161 | 1587920953716 | 0 | 1587921512 | 0 | 2 | 2 | 0 | 25 | 1300 | 22 | 5 | 1004 | 0 | 1587920944107 | 0 |  |

| id | guid | mid | mod | usn | tags | flds | sfld | csum | flags | data |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1510928805161 | G4gNm4/\*oC | 1481327979421 | 1587918621 | 0 |  LightYear::Pulmn::Intro::PulmonaryCirculation aaaa alacranlyadded bab bab::pulmonary::introduction::pulmonarycirculation  | What is the gold standard method for diagnosing pulmonary hypertension?Heart catheterization&lt;div&gt;&lt;br /&gt;&lt;/div&gt;&lt;div&gt;&lt;i&gt;The non-invasive method is via echocardiography&lt;/i&gt;&lt;/div&gt; | What is the gold standard method for diagnosing pulmonary hypertension? | 2317416375 | 0 |  |
</details>

```
INSERT INTO cards (id, nid, did, ord, mod, usn, type, queue, due, ivl, factor, reps, lapses, "left", odue, odid, flags, data) VALUES (1510928829863, 1510928805161, 1587920953716, 0, 1587921512, 0, 2, 2, 0, 25, 1300, 22, 5, 1004, 0, 1587920944107, 0, '');

INSERT INTO notes (id, guid, mid, mod, usn, tags, flds, sfld, csum, flags, data) VALUES (1510928805161, 'G4gNm4/*oC', 1481327979421, 1587918621, 0, ' LightYear::Pulmn::Intro::PulmonaryCirculation aaaa alacranlyadded bab bab::pulmonary::introduction::pulmonarycirculation ', 'What is the gold standard method for diagnosing pulmonary hypertension?Heart catheterization<div><br /></div><div><i>The non-invasive method is via echocardiography</i></div>', 'What is the gold standard method for diagnosing pulmonary hypertension?', 2317416375, 0, '');

INSERT INTO MY_TABLE(conf, dconf, decks) VALUES ('{"activeDecks": [1587920953716], "curDeck": 1587920953716, "newSpread": 0, "collapseTime": 1800, "timeLim": 0, "estTimes": true, "dueCounts": true, "curModel": "1587920938249", "nextPos": 0, "sortType": "noteFld", "sortBackwards": false, "addToCur": true, "dayLearnFirst": false, "schedVer": 2, "newBury": true, "imgocc": {"version": 1.25, "ofill": "FFEBA2", "qfill": "FF7E7E", "scol": "2D2D2D", "swidth": 3, "font": "Arial", "fsize": 24, "skip": ["Extra 1", "Extra 2"], "flds": {"id": "ID (hidden)", "hd": "Header", "im": "Image", "ft": "Footer", "rk": "Remarks", "sc": "Sources", "e1": "Extra 1", "e2": "Extra 2", "qm": "Question Mask", "am": "Answer Mask", "om": "Original Mask"}}, "localOffset": -60, "rollover": 4, "lastUnburied": 1}', '{"1": {"name": "Default", "new": {"delays": [1, 10], "ints": [1, 4, 7], "initialFactor": 2500, "separate": true, "order": 1, "perDay": 20, "bury": false}, "lapse": {"delays": [10], "mult": 0, "minInt": 1, "leechFails": 8, "leechAction": 0}, "rev": {"perDay": 200, "ease4": 1.3, "fuzz": 0.05, "minSpace": 1, "ivlFct": 1, "maxIvl": 36500, "bury": false, "hardFactor": 1.2}, "maxTaken": 60, "timer": 0, "autoplay": true, "replayq": true, "mod": 0, "usn": 0, "id": 1, "dyn": false}, "1587144167603": {"autoAgain": 120, "dyn": false, "autoplay": true, "new": {"bury": false, "ints": [15, 60, 7], "separate": true, "perDay": 0, "delays": [15, 20, 1440, 4320, 8640, 21600], "initialFactor": 2500, "order": 1}, "timer": 1, "currentValue": 120, "autoAnswer": 20, "id": 1587144167603, "maxTaken": 180, "lapse": {"leechFails": 8, "delays": [20], "mult": 0.7, "leechAction": 1, "minInt": 2}, "name": "Poorretention", "rev": {"bury": false, "maxIvl": 180, "ease4": 1.5, "perDay": 9999, "hardFactor": 1.2, "minSpace": 1, "fuzz": 0.05, "ivlFct": 0.54}, "maxLife": 120, "mod": 1587920944, "usn": 0, "replayq": true, "autoAlert": 10, "recover": 5, "anotherRetreat": false, "enableDamage": false, "damage": 5}}', '{"1": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "id": 1, "name": "Default", "mod": 1587920915}, "1587920944097": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": true, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention", "id": 1587920944097, "mod": 1587920947}, "1587920944098": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear", "id": 1587920944098, "mod": 1587920944}, "1587920944099": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::renal", "id": 1587920944099, "mod": 1587920944}, "1587920944100": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::renal::other", "id": 1587920944100, "mod": 1587920944}, "1587920944101": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::renal::other::renalbladdertumors", "id": 1587920944101, "mod": 1587920944}, "1587920944102": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::psychiatry", "id": 1587920944102, "mod": 1587920944}, "1587920944103": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::psychiatry::pharmacology", "id": 1587920944103, "mod": 1587920944}, "1587920944104": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::psychiatry::pharmacology::antidepressants", "id": 1587920944104, "mod": 1587920944}, "1587920944105": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary", "id": 1587920944105, "mod": 1587920944}, "1587920944106": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary::introduction", "id": 1587920944106, "mod": 1587920944}, "1587920944107": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary::introduction::pulmonarycirculation", "id": 1587920944107, "mod": 1587920944}, "1587920944108": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary::obstructiverestrictive", "id": 1587920944108, "mod": 1587920944}, "1587920944109": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary::obstructiverestrictive::obstructive", "id": 1587920944109, "mod": 1587920944}, "1587920944110": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::pulmonary::obstructiverestrictive::restrictive", "id": 1587920944110, "mod": 1587920944}, "1587920944111": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::reproductive", "id": 1587920944111, "mod": 1587920944}, "1587920944112": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::reproductive::embryology", "id": 1587920944112, "mod": 1587920944}, "1587920944113": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 0], "conf": 1587144167603, "usn": 0, "desc": "", "dyn": 0, "collapsed": false, "extendNew": 10, "extendRev": 50, "name": "DecreasedIntervalPoorRetention::Lightyear::reproductive::embryology::teratogensone", "id": 1587920944113, "mod": 1587920944}, "1587920953716": {"newToday": [0, 0], "revToday": [0, 0], "lrnToday": [0, 0], "timeToday": [0, 15975], "collapsed": false, "dyn": 1, "desc": "", "usn": 0, "delays": null, "separate": true, "terms": [["is:due prop:due>-1", 100, 0]], "resched": true, "return": true, "previewDelay": 10, "name": "Filtered Deck 1", "id": 1587920953716, "mod": 1587921512}}');
```

1. Move card into a Filtered deck via search on "gold".
2. Press "Again"

###### Expected Result

From Anki Desktop
* 4 Ease Options: <20m, 30m, 17d, 18d

###### Actual Result

From AnkiDroid
* 4 Ease Options:  <15m, <18m, <20m, 17d

###### Research
[x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
[x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
[x] I have searched for similar existing issues here and on the user forum

----

Anki Desktop dump:

```
>>> pp(self.col.sched.nextIvl(self.reviewer.card, 1))
... pp(self.col.sched.nextIvl(self.reviewer.card, 2))
... pp(self.col.sched.nextIvl(self.reviewer.card, 3))
... pp(self.col.sched.nextIvl(self.reviewer.card, 4))
pp(self.reviewer.card)
1200
1800
1468800
1555200
{'data': '', 'did': 1587926456564, 'due': 1587928085, 'factor': 1300, 'flags': 0, 'id': 1510928829863, 'ivl': 17, 'lapses': 6, 'left': 1001, 'mod': 1587926628, 'nid': 1510928805161, 'odid': 1587920944107, 'odue': 0, 'ord': 0, 'queue': 1, 'reps': 23, 'type': 3, 'usn': -1}
```

**Comment by mikehardy** [comment]
Reproducible upstream or no?

**Comment by david-allison-1** [comment]
Expected is the result from Anki 2.1.22

**Comment by david-allison-1** [comment]
Hmm... this is the result of a bug (we select the wrong delays for lapses), but changes upstream have also occurred (previously, Easy would have been 17 days, now it's 18). 

I've got a test in, but this will require more changes than I'd have liked to the scheduler in order to bring us closer to upstream.

Anki Change
https://github.com/ankitects/anki/blob/8dfd362fedb8fdc277eb50ff28871778f8c83257/pylib/anki/schedv2.py#L733-L738

Our bug
https://github.com/ankidroid/Anki-Android/blob/1f313a96bc786465c4303fabf62890809e9f2c21/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java#L1767

**Comment by david-allison-1** [comment]
Missed the fact this also needs a database check

**Comment by david-allison-1** [comment]
TL;DR: I'm looking at fixing a bug by graduating certain lapsed cards to "review", is this a bad/lazy idea, or is it unlikely to cause problems?

I'm close to being out of my depth with the `fixIntegrity` check.

I was planning on moving the lapse cards to a similar step (if you're 2 away from the start, move to 2 away from the start), but there's a lot of edge cases to account for, and it's unlikely there'll be a 1:1 mapping between **new** and **lapse** steps, so I'm not sure if it adds much value.

I'm getting closer to: "move all lapse cards with invalid steps back into **review**", it's much easier to confirm correctness for, but but it feels like a cop-out, and  I'm not sure if the behaviour would annoy users.  

I'm wondering whether there are common use cases where there are large steps in the lapse config, and where skipping these and graduating early would cause problems.

**Comment by mikehardy** [comment]
Only @dae can answer your questions and arrive at a tasteful solution I think. I'm never bothered enough to care about the edge cases in my own reviews so I don't know/care enough and I'm not a help

**Comment by dae** [comment]
I haven't been following along with this issue, but if the cards have invalid relearning steps, moving them back to review sounds fine - it's what would happen if they were in relearning when moved into/out of a filtered deck in the V1 scheduler.

**Comment by david-allison-1** [comment]
I've had a second look, and feel I'm missing an anti-corruption routine for `left`. There's a couple of other places where an value larger than `delays` could occur:

* When changing deck to a deck with less `delays` steps (seems the same in Anki): https://github.com/ankitects/anki/blob/961833a1374e9224e679070149984cbd8fba2a65/qt/aqt/browser.py#L1642-L1683
* When options are changed, removing steps (seems the same in Anki): https://github.com/ankitects/anki/blob/961833a1374e9224e679070149984cbd8fba2a65/qt/aqt/deckconf.py#L259

I feel I'm making an oversight as to why these don't cause excessive steps required in `(Re)Learn`

@dae Is my analysis lacking? If not, would you consider a patch to fix erroneous 'left' values, either in `card` or `sched(v2)`?

**Comment by dae** [comment]
The latter comes up from time to time, and the answer up until now has been "if you change the steps while cards are in learning, things may behave strangely until you graduate the cards or reset them". It's something I've been putting off until I have a chance to refactor the code, as it's a bit of a tangled mess at the moment, but if you want to tackle it sooner, I'd be happy to accept a PR. Probably best handled in the scheduler when the learning step is changed, rather than trying to fix it on card load or a DB check.

**Comment by mikehardy** [comment]
@david-allison-1 I think this is closable now?

**Comment by david-allison-1** [comment]
Let's close, will shelve the database check, since this happens in a few other places and isn't fixed by a check.
